api/server.js

```diff
@@ -527,69 +530,91 @@ app.postJson('/api/v1/users/:username', function(req, res) {
-          code: 401
-        }, req, res);
+  const hasFullPermission = () =>
+    new Promise((resolve, reject) => auth.check(req, 'can_update_users', null, err => {
+      if (err && err.code === 403) {
+        resolve(false);
+      } else if (err) {
+        reject(err);
+      } else {
+        resolve(true);
       }
-
-      // We're updating ourselves, but if we are updating our password we need
-      // to provide Basic Auth credentials to indicate we're entering the password
-      // fresh as of this connection
-      if (Object.keys(req.body).includes('password')) {
-        let credentials;
-        try {
-          credentials = auth.basicAuthCredentials(req);
-        } catch (err) {
-          err.statusCode = 401;
-          return serverUtils.error(err, req, res);
-        }
-
-        if (username !== credentials.username) {
-          // Make sure the Basic Auth credentials are actually for the same person
-          return serverUtils.error({
-            message: 'You do not have permissions to modify this person',
-            code: 401
-          }, req, res);
-        }
-
+    }));
+  const isUpdatingSelf = () =>
+    new Promise((resolve, reject) => auth.getUserCtx(req, (err, userCtx) => {
+      if (err) {
+        reject(err);
+      } else if (userCtx.name === username) {
+        resolve(true);
+      } else {
+        resolve(false);
+      }
+    }));
+  const basicAuthValid = () =>
+    new Promise((resolve, reject) => {
+      const credentials = auth.basicAuthCredentials(req);
+
+      if (!credentials) {
+        resolve(null); // Not attempting basic auth
+      } else if (username !== credentials.username) {
+        // Make sure the Basic Auth credentials are actually for the same person
+        reject({
+          message: 'Basic Auth user does not match the passed username',
+          code: 403
+        });
+      } else {
         auth.validateBasicAuth(credentials, err => {
           if (err) {
             console.error(`Invalid authorization attempt on /api/v1/users/${username}`);
             console.error(err);
-            return serverUtils.error({
-              message: 'Invalid Basic Auth password',
-              code: 401
-            }, req, res);
+            resolve(false); // Incorrect basic auth
+          } else {
+            resolve(true); // Correct basic auth
           }
-
-          // We have valid basic auth and can update ourselves
-          return updateUser(false);
         });
-      } else {
-        // Not trying to update password
-        return updateUser(false);
       }
     });
-  });
+  const isChangingPassword = () => Object.keys(req.body).includes('password');
+
+  // This logic is complicated but self-documenting (hopefully!)
+  Promise.all([hasFullPermission(), isUpdatingSelf(), basicAuthValid(), isChangingPassword()])
+    .then(([fullPermission, updatingSelf, basic, changingPassword]) => {
```

Comment:
This next section is really important, and should be very clear. I debated how to write this, as there are a few different approaches (nested ifs, having each line contain the whole piece of logic instead of presuming lines exist above you etc), and want to make sure this makes sense to you.